mitiq/calibration/calibration.py

```diff
@@ -0,0 +1,204 @@
+# Copyright (C) 2021 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from collections import Counter
+from math import prod, sqrt
+from typing import Any, Callable, Union, cast
+
+import cirq
+
+from mitiq import QPROGRAM, Executor, MeasurementResult, QuantumResult
+from mitiq.calibration.settings import Settings
+
+
+class Calibrator:
+    """A calibration object which keeps track of, and aids in Error
+    Mitigation parameter tuning."""
+
+    def __init__(
+        self,
+        executor: Union[Executor, Callable[[QPROGRAM], QuantumResult]],
+        settings: Settings,
+        ideal_executor: Union[
+            Executor, Callable[[QPROGRAM], QuantumResult], None
+        ] = None,
+    ):
+
+        self.executor = (
+            Executor(executor)
+            if not isinstance(executor, Executor)
+            else executor
+        )
+        self.ideal_executor = (
+            Executor(ideal_executor)
+            if ideal_executor and not isinstance(ideal_executor, Executor)
+            else ideal_executor
+        )
+        self.settings = settings
+        self.circuits = self.settings.make_circuits()
+        self.results: list[dict[str, Any]] = []
+
+    def get_cost(self) -> dict[str, int]:
+        """Returns the expected number of noisy expectation values required
+        for calibration. If an ideal_executor was used in specifying the
+        Calibrator object, the number of classical simulations is also
+        returned."""
+        num_circuits = len(self.circuits)
+        num_methods = len(self.settings.techniques)
+        num_options = prod(map(len, self.settings.technique_params.values()))
+
+        noisy = num_circuits * num_methods * num_options
+        ideal = noisy if self.ideal_executor else 0
+        return {
+            "noisy_executions": noisy,
+            "ideal_executions": ideal,
+        }
+
+    def run_circuits(self) -> list[dict[str, Any]]:
+        """Run the calibration circuits and store the ideal and noisy
+        expectation values for each circuit in `self.results`."""
+        expvals = []
+        for circuit_data in self.circuits:
+            circuit = circuit_data.circuit
+            circuit.append(cirq.measure(circuit.all_qubits()))
+
+            noisy_results = cast(
+                MeasurementResult, self.executor._run([circuit])[0]
```

Comment:
```suggestion
                MeasurementResult, self.executor._run([circuit])[0]
```
Let's remember to make an issue for upgrading `_run` to become a public method of the `Executor` class. It is strange that there is not a public way of getting bitstrings. Not urgent.

Reply:
Thanks for the reminder. Done in https://github.com/unitaryfund/mitiq/issues/1639.